README.md
# Lyft Android SDK

The Official Lyft Android SDK makes it easy to integrate Lyft into your app. More specifically, it provides:
- An easily configurable Lyft button which can display cost, ETA, and ride type. Tapping the button deeplinks into the Lyft app with pre-filled pickup/destination/ridetype.
- A Java interface for making sync/async calls to Lyft's REST APIs
- Two sample Android Activities that show how to use the SDK components.

## Registration
- You must first create a Lyft Developer account [here](https://www.lyft.com/developers).
- Once registered, you will be assigned a Client ID and will be able to generate Client Tokens.

## Setup and Installation

### Gradle:

```gradle
repositories {
  mavenCentral() // or jcenter()
}

dependencies {
    compile 'com.lyft:lyft-android-sdk:2.0.2'
}
```

If you only want to use the [Lyft API Wrapper](https://github.com/lyft/lyft-android-sdk#lyft-api-wrapper) or [Deeplink](https://github.com/lyft/lyft-android-sdk#deeplinking) portion of the SDK, you can pull them individually.
```gradle
compile 'com.lyft:lyft-android-sdk-networking:2.0.2'  // Lyft API Wrapper
compile 'com.lyft:lyft-android-sdk-deeplink:2.0.2'    // Deeplink
```

### Maven:
```xml
<dependency>
  <groupId>com.lyft</groupId>
  <artifactId>lyft-android-sdk</artifactId>
  <version>2.0.2</version>
  <type>aar</type>
</dependency>
```

## Lyft Button

Adding the Lyft Button in an XML layout is as simple as:
```xml
<com.lyft.lyftbutton.LyftButton
    android:id="@+id/lyft_button"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
    lyft:lyftStyle="lyftMulberryDark"
    />
```
We recommend setting the width/height to `wrap_content`, which results in a width of 260dp and a height of 50dp.
Otherwise, please keep in mind that a smaller width/height may result in undesirable UI, such as overlapping text.

To load ETA/cost:
```java
ApiConfig apiConfig = new ApiConfig.Builder()
        .setClientId("your_client_id")
        .setClientToken("your_client_token")
        .build();

LyftButton lyftButton = (LyftButton) findViewById(R.id.lyft_button);
lyftButton.setApiConfig(apiConfig);

RideParams.Builder rideParamsBuilder = new RideParams.Builder()
        .setPickupLocation(37.7766048, -122.3943576)
        .setDropoffLocation(37.759234, -122.4135125);
rideParamsBuilder.setRideTypeEnum(RideTypeEnum.STANDARD);

lyftButton.setRideParams(rideParamsBuilder.build());
lyftButton.load();
```

Addresses can also be used for pickup/dropoff locations. Each address results in an extra API call to obtain the corresponding lat/lng pair. Therefore, we recommend using lat/lng values directly if they are available. 
```java
RideParams.Builder rideParamsBuilder = new RideParams.Builder()
        .setPickupAddress("185 Berry St, San Francisco, CA 94107")
        .setDropoffAddress("2300 Harrison St, San Francisco, CA 94110");
```

### Ride types
Lyft is growing very quickly and is currently available in [these cities](https://www.lyft.com/cities). Please keep in mind that some ride types (such as Lyft Line) are not yet available in all Lyft cities. If you set the ride type of the button  and it happens to be unavailable, the button will default to the Lyft Standard ride type. You can utilize the [`/v1/ridetypes`](https://developer.lyft.com/docs/availability-ride-types) endpoint to get a list of the available ride types in an area.

### Button styles
To specify the button style via XML, use the `lyft:lyftStyle` attribute or set it programmatically:
```java
lyftButton.setStyle(LyftStyle.MULBERRY_DARK);
```

There are 5 styles to pick from:

![lyft-styles](https://cloud.githubusercontent.com/assets/13209348/17683300/88f86446-6306-11e6-81e6-bc42fc77650e.png)

## Lyft API Wrapper
The SDK provides wrapper methods around Lyft's REST APIs - this can be helpful when you want to build a more custom integration with the Lyft platform vs making HTTP requests directly.

The SDK uses Square's [Retrofit 2](http://square.github.io/retrofit/) networking library. For each Lyft API endpoint, there is a corresponding Java method. The return type is a `Call` object which can be executed synchronously or asynchronously. See [LyftApi.java](https://github.com/lyft/lyft-android-sdk/blob/master/networking/src/main/java/com/lyft/networking/apis/LyftApi.java) for all the API methods.

```java
LyftApi lyftPublicApi = new LyftApiFactory(apiConfig).getLyftApi();
Call<EtaEstimateResponse> etaCall = lyftPublicApi.getEtas(37.7766048, -122.3943576, "lyft");
```
Asynchronous:
```java
etaCall.enqueue(new Callback<EtaEstimateResponse>() {
    @Override
    public void onResponse(Call<EtaEstimateResponse> call, Response<EtaEstimateResponse> response) {
        EtaEstimateResponse etaEstimateResponse = response.body();
        Eta eta = etaEstimateResponse.eta_estimates.get(0);
    }

    @Override
    public void onFailure(Call<EtaEstimateResponse> call, Throwable t) {
        Log.d("MyApp", t.toString());
    }
});
```

Synchronous:
```java
EtaEstimateResponse etaEstimateResponse = etaCall.execute().body();
Eta eta = etaEstimateResponse.eta_estimates.get(0);
```

### RxJava Observables
If you already use RxJava (your app must have it as a dependency), then you can obtain an `Observable` instead of a `Call` object. Simply use [LyftApiRx](https://github.com/lyft/lyft-android-sdk/blob/master/networking/src/main/java/com/lyft/networking/apis/LyftApiRx.java) instead of LyftApi.

```java
LyftPublicApiRx lyftPublicApiRx = new LyftApiFactory(apiConfig).getLyftPublicApiRx();
Observable<EtaEstimateResponse> etaObservable = lyftPublicApiRx.getEtas(37.7766048, -122.3943576, "lyft");
```

### ProGuard
If you are directly using the Lyft Wrapper API without using the `Lyft Button`, then you may need to add the the ProGuard rules for Retrofit. Please see [proguard-rules.pro](https://github.com/lyft/lyft-android-sdk/blob/master/lyft-button/proguard-rules.pro). This only applies if you are using ProGuard.

## Deeplinking
The SDK provides direct [deeplinking](https://developer.lyft.com/docs/deeplinking) to the Lyft app for those developers who prefer to handle their own custom deeplinking vs relying on the Lyft Button.. The `deeplink` module of the SDK includes this logic and makes it easy to launch the Lyft app.
```java
DeepLinkParams deepLinkParams = new DeepLinkParams.Builder()
        .setClientId("your_client_id")
        .setRideType(RideTypeEnum.SHARED)
        .setPickupLocation(37.7766048, -122.3943576)
        .setDropoffLocation(37.759234, -122.4135125)
        .build();

DeepLink.launchLyftApp(getContext(), deepLinkParams);
```

## Sample Activities
Checkout the sample activites which are included in the SDK, in the `sample-app` module:
  - `SampleBasicActivity`: Includes minimal code to set up the Lyft Button.
  - `SampleLocationAwareActivity`: Gets the device's current GPS location and calls the `/v1/ridetypes` endpoint to get a list of the available ride types in the area (i.e. Lyft Line, Lyft Plus, etc). The user is able to select an available ride type via a dropdown. The Lyft Button then displays the current ETA and cost of that ride type to a specified destination.

You can specify which activity to launch in the `sample-app`'s [AndroidManifest.xml](https://github.com/lyft/lyft-android-sdk/blob/master/sample-app/src/main/AndroidManifest.xml).

## Support

If you're looking for help configuring or using the SDK, or if you have general questions related to our APIs, the Lyft Developer Platform team provides support through our [forum](https://developer.lyft.com/discuss) as well as on Stack Overflow (using the `lyft-api` tag)

## Reporting security vulnerabilities

If you've found a vulnerability or a potential vulnerability in the Lyft Android SDK,
please let us know at security@lyft.com. We'll send a confirmation email to
acknowledge your report, and we'll send an additional email when we've
identified the issue positively or negatively.


CHANGELOG.md
Version 2.0.2 *(2020-09-20)*
----------------------------
* Fixed a bug that caused an NPE when attempting to display a price on the Lyft Button.

Version 2.0.1 *(2019-11-18)*
----------------------------
* Fixed a bug that prevented promo codes from being applied to the Lyft Button.

Version 2.0.0 *(2019-07-12)*
----------------------------
* Update Gradle version from `2.14.1` to `4.10.1`, and Android Gradle Plugin from `2.2.2` to `3.3.2`
* The compile SDK target for Android has been bumped to API 28.
* The following dependencies have been updated:
  * Android Build Tools `23.0.3` -> `28.0.3`
  * Android Support Libraries `23.4.0` -> `28.0.0`
  * Google Play Services `9.4.0` -> `17.0.0`
  * Mockito `1.10.19` -> `2.25.0`
  * Retrofit `2.1.0` -> `2.6.0`
* `LyftUserApi`, `LyftUserApiRx` and all endpoints that require a Lyft User OAuth token **have been removed**. External clients will no longer have access to features that require user information.
* `LyftPublicApi` and `LyftPublicApiRx` have been renamed to `LyftApi` and `LyftApiRx` respectively, to reflect new changes regarding the UserApi client.
* The RxJava call adapter for Retrofit has been updated to RxJava2. Users of the RxJava Retrofit interface will have to update their codebase to take RxJava2 Observables.
* All models returned by network calls have been documented with JetBrain `@Nullable` and `@NotNull` annotations.
* `LyftApiFactory` now by default generates a `LyftApi` that **throws a `PartialResponseException`** when the server response doesn't abide by the `@Nullable` and `@NotNull` annotations of the parsed model. The client must be able to handle a `PartialResponseException` when calling for Lyft endpoints.
  * If the client wishes to handle null checking manually, call `LyftApiFactory#getUncheckedLyftApi` to get a Retrofit client that does not massage responses.
* `RideTypeEnum` has been moved from the Lyft Button module to the Deeplink module, with the following changes:
  * `toString()` is no longer valid when fetching for a ride type key. Use `RideTypeEnum#getRideTypeKey()` instead.
  * `getDisplayName()` is still valid for fetching human readable text for each ride type.
* `RideTypeEnum` has been updated to reflect the current ride offerings  provided by Lyft:
  * `LINE` has been replaced by `SHARED`
  * `CLASSIC` has been replaced by `STANDARD`
  * `PLUS` has been replaced by `XL`
  * `LUX`, `LUX_BLACK`, `LUX_BLACK_XL` has been added into `RideTypeEnum`
  * `null` can be used in place of `RideTypeEnum.ALL` in existing usages 
* `DeepLinkParams` and its builder now takes in `RideTypeEnum` to specify a ride type, instead of a `String`.

Version 1.0.3 *(2017-02-21)*
----------------------------
Rename couponCode to promoCode
Add promoCode to RideParams
Make ResultLoadedCallback interface public

Version 1.0.2 *(2016-11-22)*
----------------------------
Coupon code parameter in ridetype deeplink
Expose address parameters in deeplink

Version 1.0.1 *(2016-11-15)*
----------------------------
Small deeplink fixes.
Avoid NPE if eta_seconds is null.

Version 1.0.0 *(2016-09-01)*
----------------------------
Initial release.


sample-app/src/main/java/com/lyft/sdk/SampleBasicActivity.java
package com.lyft.sdk;

import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;

import com.lyft.deeplink.RideTypeEnum;
import com.lyft.lyftbutton.LyftButton;
import com.lyft.lyftbutton.RideParams;
import com.lyft.networking.ApiConfig;
import com.lyft.sdk.testing.TestLyftApiFactory;
import com.lyft.sdk.testing.ui.SpinnerPairItem;

import java.util.Arrays;
import java.util.List;

public class SampleBasicActivity extends Activity {

    private Spinner rideTypeSpinner;
    private LyftButton lyftButton;

    private List<SpinnerPairItem<RideTypeEnum>> spinnerList = Arrays.asList(
            new SpinnerPairItem<>(RideTypeEnum.SHARED.getDisplayName(), RideTypeEnum.SHARED),
            new SpinnerPairItem<>(RideTypeEnum.STANDARD.getDisplayName(), RideTypeEnum.STANDARD),
            new SpinnerPairItem<>(RideTypeEnum.XL.getDisplayName(), RideTypeEnum.XL),
            new SpinnerPairItem<>(RideTypeEnum.LUX.getDisplayName(), RideTypeEnum.LUX),
            new SpinnerPairItem<>(RideTypeEnum.LUX_BLACK.getDisplayName(), RideTypeEnum.LUX_BLACK),
            new SpinnerPairItem<>(RideTypeEnum.LUX_BLACK_XL.getDisplayName(), RideTypeEnum.LUX_BLACK_XL)
    );

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_sample_basic);
        setupSpinner();

        ApiConfig apiConfig = new ApiConfig.Builder()
                .setClientId("your_client_id")
                .setClientToken("your_client_token")
                .build();

        lyftButton = findViewById(R.id.lyft_button);
        lyftButton.setApiConfig(apiConfig);

        loadButtonWithRideType(RideTypeEnum.STANDARD);
    }

    private void setupSpinner() {
        rideTypeSpinner = findViewById(R.id.sample_basic_activity_spinner);

        ArrayAdapter<SpinnerPairItem<RideTypeEnum>> adapter = new ArrayAdapter<>(
                this,
                android.R.layout.simple_spinner_item,
                spinnerList);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        rideTypeSpinner.setAdapter(adapter);
        rideTypeSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                loadButtonWithRideType(spinnerList.get(position).getData());
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                //Do nothing
            }
        });
    }

    private void loadButtonWithRideType(RideTypeEnum rideTypeEnum) {
        RideParams.Builder rideParamsBuilder = new RideParams.Builder()
                .setPickupLocation(37.7766048, -122.3943629)
                .setDropoffLocation(37.759234, -122.4135125);
        rideParamsBuilder.setRideTypeEnum(rideTypeEnum);

        lyftButton.setRideParams(rideParamsBuilder.build());
        lyftButton.load();
    }
}
